docs/design.md

```diff
@@ -0,0 +1,371 @@
+
+   // private
+   InsertFullIndex(key shard.Key, index FullIndex) error
+   DeleteFullIndex(key shard.Key) (bool, error)
+}
+
+type ManifestRepo interface {
+   // public
+   GetManifest(key ManifestKey) (Manifest, error)
+   InsertManifest(key ManifestKey, manifest Manifest) error
+   DeleteManifest(key ManifestKey) (bool, error)
+   ExistsManifest(key ManifestKey) (bool, error)
+}
+
+type ManifestKey struct {
+   Shard    key.Shard
+   Rule     string
+   Version  string
+}
+
+type FullIndex interface {
+   Offset(c cid.Cid) (offset int64, err error) // inexistent: offset=-1, err=nil
+   Contains(c cid.Cid) (bool, error)
+   Len() (l int64, err error)
+   ForEach(func(c cid.Cid, offset int64) (ok bool, err error)) error
+}
+```
+
+## DAG access layer
+
+This layer is responsible for serving DAGs or sub-DAGs from one or many shards.
+Initially, queries will require the shard key. That is, queries will be bounded
+to a single identifiable shard.
+
+In the future, when the cross-shard top-level index is implemented, the DAG
+store will be capable of resolving the shard for any given root CID.
+
+DAG access layer allows various access abstractions:
+   1. Obtaining a Blockstore bound to a single shard.
+   2. Obtaining a Blockstore bound to a specific set of shards.
+   3. Obtaining a global Blockstore.
+   4. `ipld.Node` -- TBD.
+
+At this point, we are only concerned with (1). The remaining access patterns
+will be elaborated on in the future.
+
+```go
+type DAGAccessor interface {
+   Shard(key shard.Key) ShardAccessor
+}
+
+type ShardAccessor interface {
+   Blockstore() (ReadBlockstore, error)
+}
+```
+
+## Requirements for CARv2
+
+- Index needs to be detachable.
+- Index offsets need to be relative to the CARv1, and not absolute in the
+  physical CARv2 file.
+- Index needs to be iterable.
+- Given any CAR file, we should be able to determine its version and
+  characteristics (concretely, indexed or not indexed).
+- Given a CARv1 file, it should be possible to generate a CARv2 index for it in
+  detached form.
+- Given a CARv2 file, we should be able to decompose it to the corresponding
+  CARv1 payload and the Carbs index. The CAR payload should be exposed with
+  `io.ReaderAt` and `io.Reader` abstractions.
+- Given an indexed CARv2 file, it should be possible to instantiate a
+  `ReadBlockstore` on it in a self-sufficient manner.
+- Given a CARv1 or unindexed CARv2, it should be possible to instantiate a
+  `ReadBlockstore` with an index provided by us.
+- It should be possible to write an indexed CARv2 in streaming fashion. The
+  CARv2 library should do the right thing depending on the declared write
+  characteristics and the output characteristics. For example:
+   1. the writer may declare that blocks are provided in depth-first order and
+      without uniqueness guarantees, but may state they want the output to be in
+      depth-first order AND deduplicated. In this case, the CARv2 library must
```

Comment:
Just want to point out that this should be _"in an order  determined by the given IPLD selector"_ here as Graphsync could be using any IPLD selector as given by the Graphsync client. 

It just so happens that the default selector we use in Filecoin Markets is the "depth-first" selector if the user dosen't mandate the use of a specific selector.

Reply:
- Also, important to note that as discussed with @willscot, the CARV2 writer wont care about the selector/block order in the first iteration. 

- The CARv2 writer assumes that the caller knows what it's doing (i.e. it's writing blocks based on an order determined by a valid IPLD selector and that the CARv2 file will form a coherent DAG in the end) and does NOT enforce any restrictions on the order of the writes.

- So really, in the first  immediate iteration, the writer will declare that there are no uniqueness guarantees and that it expects the output to be de-duplicated but will make no declarations about the selector /block order as the CARv2 writer does NOT care about it.

Reply:
@willscott Just to be clear, the first iteration _will_ deduplicate blocks but will not reorder them (as we can make the hard guarantee that blocks are delivered in the order they are to be written), right?

Reply:
The current carbon will write a block everytime you call `Put`, so won't automatically dedup for you.
Does it make more sense for you as the caller to do the `Has` check to skip duplicates, or do we want that logic in the library?

@masih - I haven't looked closely at the method for opening the carv2 as a blockstore. Do we anticipate there'll be a reasonably ergonomic place to pass in an option for deduplicating puts when we do that?

I don't think this is hard, it's just a matter of where we do it.

Reply:
Can you clarify "depth-first" please? This could mean that visited blocks are yielded before or after visiting child blocks. We need to be clear about this point if we're baking this in anywhere and not just relying on selectors to do this for us.